types/src/keyless/proof_simulation.rs

```diff
@@ -0,0 +1,258 @@
+
+        Ok((pk, vk))
+    }
+
+    /// Generates a random common reference string for
+    /// a circuit using the provided R1CS-to-QAP reduction.
+    #[inline]
+    pub fn generate_random_parameters_and_trapdoor_with_reduction<C>(
+        circuit: C,
+        rng: &mut impl Rng,
+    ) -> R1CSResult<(ProvingKeyWithTrapdoor<E>, VerifyingKey<E>)>
+    where
+        C: ConstraintSynthesizer<E::ScalarField>,
+    {
+        let alpha = E::ScalarField::rand(rng);
+        let beta = E::ScalarField::rand(rng);
+        let gamma = E::ScalarField::rand(rng);
+        let delta = E::ScalarField::rand(rng);
+
+        let g1_generator = E::G1::rand(rng);
+        let g2_generator = E::G2::rand(rng);
+
+        let pk = Groth16::<E,QAP>::generate_parameters_with_qap(
+            circuit,
+            alpha,
+            beta,
+            gamma,
+            delta,
+            g1_generator,
+            g2_generator,
+            rng,
+        ).unwrap();
+
+        Ok((
+        ProvingKeyWithTrapdoor {
+            gamma_abc_g1: pk.vk.gamma_abc_g1.clone(),
+            alpha,
+            beta,
+            delta,
+            gamma,
+            g1: g1_generator.into_affine(),
+            g2: g2_generator.into_affine(),
+        },
+        pk.vk.clone()))
+    }
+
+    /// Create a Groth16 proof that is zero-knowledge using the provided
+    /// R1CS-to-QAP reduction.
+    /// This method samples randomness for zero knowledges via `rng`.
+    #[inline]
+    pub fn create_random_proof_with_trapdoor(
+        public_inputs: &[E::ScalarField],
+        pk: &ProvingKeyWithTrapdoor<E>,
+        rng: &mut impl Rng,
+    ) -> R1CSResult<Proof<E>>
+    where
+    {
+        let a = E::ScalarField::rand(rng);
+        let b = E::ScalarField::rand(rng);
+
+        Self::create_proof_with_trapdoor(pk, a, b, public_inputs)
+    }
+
+    /// Creates proof using the trapdoor
+    #[cfg(test)]
+    pub fn create_proof_with_trapdoor(
+        pk: &ProvingKeyWithTrapdoor<E>,
+        a: E::ScalarField,
+        b: E::ScalarField,
+        input_assignment: &[E::ScalarField],
+    ) -> R1CSResult<Proof<E>> {
+        let public_inputs = input_assignment;
+        let mut g_ic = pk.gamma_abc_g1[0].into_group();
+        for (i, b) in public_inputs.iter().zip(pk.gamma_abc_g1.iter().skip(1)) {
+            g_ic.add_assign(&b.mul_bigint(i.into_bigint()));
+        }
+        g_ic = g_ic * pk.gamma;
+
+        let delta_inverse = pk.delta.inverse().unwrap();
+        let ab = a * b;
+        let alpha_beta = pk.alpha * pk.beta;
+
+        let g1_ab = pk.g1 * ab;
+        let g1_alpha_beta = pk.g1 * alpha_beta;
+        
+        let g1_a = pk.g1 * a;
+        let g2_b = pk.g2 * b;
+
+        let g1_c = (g1_ab - g1_alpha_beta - g_ic) * delta_inverse; 
+
+        Ok(Proof {
+            a: g1_a.into_affine(),
+            b: g2_b.into_affine(),
+            c: g1_c.into_affine(),
+        })
+    }
+}
+
+use ark_circom::CircomConfig;
```

Comment:
Can you move these to the top of the file?

Reply:
Yes done